### PR TITLE
Use Github Actions for deployment

### PR DESCRIPTION
**Pull request by tobiasdiez**
Use Github actions and jpackager for deployment. Rip circleci ❤️ . Rip install4j❤️

The following files are created and deployed to builds.jabref.org (and as github artifacts):
- for Windows: JabRef-5.0.0.msi (installer) and JabRef-portable-windows.zip (unzip & run)
- for Linux: jabRef-5.0.0-1_amd64.deb & jabRef-5.0.0-1_x64.rpm (packages) and JabRef-portable-linux.tar.gz (untar & run)
- for MacOS: JabRef-5.0-dev.dmg & JabRef-5.0-dev.pck (packages) and JabRef-portable-macos.tar.gz (untar & run) 

Thus, we can provide an installer (or package) as well as a runable version for each OS.

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Related: https://help.github.com/en/articles/software-in-virtual-environments-for-github-actions

**Comment by tobiasdiez** [comment]
So this is working now and produces installers / application packages for windows, linux and macos. It would be nice if you could try out the build from https://github.com/JabRef/jabref/pull/5312/checks?check_run_id=221671677 (right top corner "Artifacts", the ones not containing "raw"). These zip files contain the packages and one folder with a 'portable' version (i.e. everything that is needed to run JabRef but no installer and not packed together).

Open things (help much appreciated!)
- [ ] Upload build artifacts to build.jabref.org (which ones?)
- [ ] Customize installers & packages. https://openjdk.java.net/jeps/343 has an overview of possible arguments (which can be added in the jpackage gradle task). I've not enough experience with linux / macos packing to be able to decide which of them are nice to have / important.

**Comment by airsquared** [comment]
@koppor I created a PR to setup-java (actions/setup-java#19) that adds support for JavaFX, so you can use that if it's merged(or you can use my fork directly).

File: .github/workflows/deployment.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        elif (url.endswith("zip")):
+          zip = zipfile.ZipFile(tmpfile)
+          zip.extractall()
+          zip.close() 
+      shell: python
+    - name: Build runtime image
+      run: ./gradlew jlinkZip
+    - name: Build installer
+      run: |
+        export BADASS_JLINK_JPACKAGE_HOME="${GITHUB_WORKSPACE}${{ matrix.jdk14Path }}"
```
Comment (stefan-kolb) [inline .github/workflows/deployment.yml]: 😸

File: build.gradle [hunk 2]
```diff
@@ -665,10 +665,16 @@ jlink {
         provides 'com.microsoft.applicationinsights.core.dependencies.io.grpc.ManagedChannelProvider' with 'com.microsoft.applicationinsights.core.dependencies.io.grpc.netty.shaded.io.grpc.netty.NettyChannelProvider'
     }
 
-    // This is experimental thing; could replace install4j
     jpackage {
-        // Download from https://jdk.java.net/jpackage/ and change path accordingly
-        jpackageHome = 'H:\\Downloading\\openjdk-13-jpackage+49_windows-x64_bin\\jdk-13'
+        // In order for this to work, you need to dowload jpackage from https://jdk.java.net/jpackage/
+        // and put the path to the jdk-14 folder in the environment variable  BADASS_JLINK_JPACKAGE_HOME
```
Comment (stefan-kolb) [inline build.gradle]: 🌵

**Comment by stefan-kolb** [comment]
600 MB? Isn't that even larger than with Install4J?

**Comment by tobiasdiez** [comment]
These are 2 installers plus one portable version. So in the end 200mb for one installer, which is approximately the same size as install4j

**Comment by stefan-kolb** [comment]
On a side note: while Github actions are very fast and convenient, the download of the aertifatcs is __very slow__ as it seems. 
Some catches:
- The icons for JabRef seem to be missing right now.
- Everything is tagged JabRef-1.0 at the moment
- As far as i remember, we had some tweaks inside Install4J for installation on admin and non-admin user accounts. We should check if the installer puts Jabref inside the correct folders
- Difference between exe and msi?
- Currently the installer is tagged as `Unknown issuer`
- No installer starts up for me

**Comment by tobiasdiez** [comment]
Thanks @stefan-kolb for your feedback. I've improved the Windows installer accordingly. The only thing that I couldn't figure out was how to replace the icons in the installer - for this I wrote an email to the java devs (apparently like they like to have feedback by email; I'm happy that they don't require postal mail).

**Comment by koppor** [comment]
- [ ] The upload misses to sort in the binaries in the correct directories.

   grafik

- [ ] The upload misses to delete the old files (of the same branch)
- [ ] For security reasons, we use `sftp` and not `scp`. Maybe, I need to reconfigure the whole server or try to find a complete new solution.
- [ ] no symlink `-latest` is generated for the latest master release. (Refs https://github.com/JabRef/jabref/issues/2953).

To work on a new server configuration, or a completely new server (refs https://github.com/koppor/jabref-admin/issues/22), and a new upload configuration can be separated from the switch to `jpacakge`, can't it? I would propose to keep the old script `scripts/upload-to-builds.jabref.org.sh`.

**Comment by tobiasdiez** [comment]
Thanks @koppor! I fixed the deployment to the correct folder so that everything should work now. 

Ready for review!

**Comment by koppor** [comment]
Upload still does not delete the old files. Maybe, I need to create a server-side script (and thus requiring more logic on the server side)

image

**Comment by koppor** [comment]
- [ ] The portable edition on linux should have the prefix `build/distribution/`. Example filename: `build/distribution/JabRef/runtime/legal/java.naming/LICENSE`
- [ ] JabRef does not start on Linux@ChromeOS - window is empty and I get following output:
    ```
    oliverkopp@penguin:/tmp/build/distribution/JabRef/bin$ ./JabRef 
    ERROR StatusLogger Log4j2 could not find a logging implementation. Please add 
    log4j-core to the classpath. Using SimpleLogger to log to the console...
    ERROR ThemeLoader Could not watch css file for changes 
    jrt:/org.jabref/org/jabref/gui/Base.css
     java.lang.UnsupportedOperationException
    ```

**Comment by tobiasdiez** [comment]
Since the build files are overwritten automatically, there shouldn't be a need to clean the folder. (Of course, right now the githubactions folder is a mess because of all the experiments.)

What do you mean with the `build/distribution` folder?

The empty window might be a result of a missing hardware acceleration on chrome os. You can try to add a vm arg to the luncher script as described here https://stackoverflow.com/a/18773007/873661. Is this a result of the packaging or do you have the same problem when running from code?

**Comment by koppor** [comment]
- [ ] Regarding the file names: I like `JabRef_windows-x32_5_0-dev--snapshot--2019-09-12--githubActions--92c970011.XXX` better than `JabRef-5.0.1.XXX`. A version string (such as `5.0.1`) implies that there is ONE binary for that version (meaning, there is a 1:1 mapping to a binary). In the current implementation, it is a 1 to n mapping. For one version, there exist multiple binaries. This is IMHO not a good user experience... Would it be possible to use `snapshot` in the filenames. It is OK for me to drop the release date from the executable (and possibly also in the info screen)

Old screen:

grafik

New screen:

grafik

I can accept that we do not include any release date and commit id into the executable any more. I think, in the last years, we never checked for a commit id.

I think, I cannot accept that we do not have `dev` or `SNAPSHOT` in the filename.

**Comment by koppor** [comment]
With `build/distribution` folder I mean that if I as user unzip/untar the binary, I get the folder `build`. In this folder, there is the folder `distribution`. Then, I find `JabRef`. On windows, I find `JabRef.exe`. On Linux, I find `bin/jabref`.

I would have exepected `JabRef/JabRef.exe`, without the additional nesting in `build/distribution`.

**Comment by tobiasdiez** [comment]
The problem was that on Windows and Linux the version string had to be of the form `5.0.0.0` without additional "dev", "alpha" or whatever specification. I didn't bother changing the file names as I don't think its that important (we always refer the user to "build.jabref.org" with the words "please try out the development version..." so it that information is there from the context). But feel free to add "dev" to the files - shouldn't be that difficult as you can simply run any bash command you want. (I don't think we need to add the date (already present as file creation date), snapshot (duplication with dev) or the branch name & commit hash (not important for the user).)

The linux file folder structure of the tar should be fixed now. 

That leaves the problem with the version string in the about dialog. I agree that it should contain all information. I didn't change anything fundamental in the script https://github.com/JabRef/jabref/blob/701196a31c7e624815b32a175049b1badb735870/build.gradle#L511-L542 but I guess it is not called before https://github.com/JabRef/jabref/blob/701196a31c7e624815b32a175049b1badb735870/build.gradle#L281. Proposal: instead of overwriting `project.version` (and having the additional `versionThreeDots`) make a function that creates a full version info and use this in the expansion of `build.properties`. But I don't have the gradle experience so I leave it to you to decide if that's the best approach.

**Comment by tobiasdiez** [comment]
Yes, its important in the about dialog (but not in the file name).

**Comment by koppor** [comment]
(Filename discussion)

Side note: In install4j, we kept the four dot version and changed the version internally only. With jpackage, however, this hack seems to stop working.

- date (already present as file creation date) --> If I download the file, the file gets a new date on my machine
- snapshot (duplication with dev) --> snapshot (more common in the java world) or dev (what we used until now) - both solutions are OK for me.
- commit hash (not important for the user) -- OK for me to have that in the about dialog only. -- We then accept that two files with the same filename (on different directories) lead to different versions (is OK as other tools like java.exe also do it)

So, OK for me to go with `5.0-dev` as filename. In case we need `5.0.0-dev`, we should adapt our program versioning to semantic versioning, too. Otherwise, it will be confusing for users, why only one dot is communicated, but the program version itself has two dots.

**Comment by koppor** [comment]
(Gradle discussion)

I need to google that. No clue.

**Comment by koppor** [comment]
Note to self: Current file list: http://builds.jabref.org/githubActions/

**Comment by tobiasdiez** [comment]
The version string in the about dialog is now working (in the end, the `-Pdev=true` switch was missing).
I'll merge this PR now to get early feedback from the users. Renaming the files can be done later.

**Comment by koppor** [comment]
We seem to switch to the [fail-fast](https://www.scrum-academy.de/scrum-lexikon/fail-fast-schnell-scheitern/) pattern... OK for me, but a different philosophy. The master branch still seems to be release-ready. 😅

**Comment by koppor** [comment]
I checked all builds at https://github.com/JabRef/jabref/runs/237858080. We do not seem to distribute a `jar` file anymore. Is this intended? 

So we break the Gentoo distribution (refs https://github.com/JabRef/jabref/issues/3380#issuecomment-341087553).

**Comment by tobiasdiez** [comment]
Yes, the jar file was not working - that was the main motivation for this PR in the first place. Thus, me merging quickly is more a fix-fast approach ;-)

The packages for Linux need to be changed indeed. But since we now provide rpm and deb packages, as well as a version where you only need to unpack it, the distribution should be more easy now.

**Comment by Siedlerchr** [comment]
There seems to be some problem with the jdk setup now.
https://github.com/JabRef/jabref/runs/240665291

**Comment by koppor** [comment]
3 hours later, in the master branch, everything was OK: https://github.com/JabRef/jabref/commit/3e1195d5d551014068a1731fa77f5f96b99a3be2/checks

Temporary issue?

**Comment by Siedlerchr** [comment]
The problem was in the jdk action, seems like they fixed it

**Pull request merged**